cpp/src/sampling/random_walks_impl.cuh

```diff
@@ -174,16 +352,203 @@ struct node2vec_selector {
+
+    if (edge_weight_view) {
+      auto [sample_offsets, sample_e_op_results] =
+        cugraph::per_v_random_select_transform_outgoing_e(
+          handle,
+          graph_view,
+          vertex_frontier.bucket(0),
+          cugraph::edge_src_dummy_property_t{}.view(),
+          cugraph::edge_dst_dummy_property_t{}.view(),
+          *edge_weight_view,
+          node2vec_random_walk_e_bias_op_t<vertex_t, weight_t, weight_t>{
+            p_, 
+            q_, 
+            raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()), 
+            raft::device_span<vertex_t const>(intersection_indices.data(), intersection_indices.size()), 
+            raft::device_span<vertex_t const>(current_vertices.data(), current_vertices.size()), 
+            raft::device_span<vertex_t const>((*previous_vertices).data(), (*previous_vertices).size())
+          },
+          cugraph::edge_src_dummy_property_t{}.view(),
+          cugraph::edge_dst_dummy_property_t{}.view(),
+          *edge_weight_view,
+          node2vec_sample_edges_op_t<vertex_t, weight_t>{},
+          rng_state_,
+          size_t{1},
+          true,
+          std::make_optional(
+            thrust::make_tuple(vertex_t{cugraph::invalid_vertex_id<vertex_t>::value}, weight_t{0.0})));
+        minors = std::move(std::get<0>(sample_e_op_results));
+        weights = std::move(std::get<1>(sample_e_op_results));
+    } else {
+      auto[ sample_offsets, sample_e_op_results] =
+      cugraph::per_v_random_select_transform_outgoing_e(
+        handle,
+        graph_view,
+        vertex_frontier.bucket(0),
+        cugraph::edge_src_dummy_property_t{}.view(),
+        cugraph::edge_dst_dummy_property_t{}.view(),
+        cugraph::edge_dummy_property_t{}.view(),
+        node2vec_random_walk_e_bias_op_t<vertex_t, weight_t, void>{
+          p_, 
+          q_, 
+          raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()), 
+          raft::device_span<vertex_t const>(intersection_indices.data(), intersection_indices.size()), 
+          raft::device_span<vertex_t const>(current_vertices.data(), current_vertices.size()), 
+          raft::device_span<vertex_t const>((*previous_vertices).data(), (*previous_vertices).size())
+        },
+        cugraph::edge_src_dummy_property_t{}.view(),
+        cugraph::edge_dst_dummy_property_t{}.view(),
+        cugraph::edge_dummy_property_t{}.view(),
+        node2vec_sample_edges_op_t<vertex_t, void>{},
+        rng_state_,
+        size_t{1},
+        true,
+        std::make_optional(vertex_t{cugraph::invalid_vertex_id<vertex_t>::value}));
+      minors = std::move(sample_e_op_results);
+    }
+
+    //  Copy current vertices to previous vertices for two-order walk
+    thrust::copy(handle.get_thrust_policy(), 
+                 current_vertices.begin(), 
+                 current_vertices.end(), 
+                 (*previous_vertices).data());
```

Comment:
You may better perform this copy outside this function. This is really not part of a single step random walk but more for preparing for the next step. And if we perform this outside this function, previous_vertices can be const, so all the input parameters are const. This makes this function more functional (all the input parameters are non-modifiable, and you return updates only through return values).

Reply:
The function call on Line 682 uses the return to overwrite `current_vertices`, this is the latest point to make the copy before the data is lost.  In order to perform the copy outside of the function, I would need to have a separate variable to temporarily store the results so that the copy could be performed, then copy the data from the temp variable to 'current_vertices'.  Would that suffice?

Reply:
What about taking (current_vertices, previous vertices) as R-values and return new (current_vertices, previous vertices)?
```
std::tuple<
rmm::device_uvector<vertex_t> /* new current vertices */,
rmm::device_uvector<vertex_t> /* new previous vertices */,
... >
follow_random_edge(...,
rmm::device_uvector<vertex_t>&& /* current_vertices */,
rmm::device_uvector<vertex_t>&& /* previous vertices */,
...) {
...
previous_vertices = std::move(current_vertices);
return std::make_tuple(std::move(minors), std::move(previous_vertices), ...);
}
```